### PR TITLE
Make full use of the registry

### PR DESCRIPTION
**Pull request by dmonzonis**
OK, this PR is gonna be kinda big, so sorry for that.

I'll try to summarize what I've done:
- Make registry use enums as keys (IDs) instead of strings.
- Make backpack store IDs instead of entity objects.
- Remove position and game_map from entity constructors. Entities should now be placed in a map with the `place` method that I've also added.
- Make level generation independent of the player actor. Now, the player actor is placed by the Dungeon class when moving from one level to another. Also, separate layout generation from entity population.
- Rename *monsters.json* to *actors.json*, in case we have NPCs with non-violent behaviors in the future.
- Make registry instantiate new Actor and Item objects when retrieving them.
- Make Behavior's `take_turn` static, so there's no more need to instantiate behavior objects anymore (is there now a reason for them to be a singleton if we're not even instantiating them once?).

### CODE REVIEW

File: backpack.py [hunk 1]
```diff
@@ -21,27 +21,31 @@ class Backpack(metaclass=Singleton):
...
-        for item in cls.contents:
-            if item.name == name:
-                return item
+    def __init__(cls, registry):
+        cls.registry = registry
 
-    def add(cls, item, qty):
+    def _search(cls, id):
+        for item_id in cls.contents:
+            if item_id == id:
```
Comment (Elkasitu) [inline backpack.py]: this can be confusing as `id` is a python built-in

File: entities.py [hunk 2]
```diff
@@ -28,38 +28,37 @@ class Entity(ABC):
...
-        game_map (Level): Reference to the level where the entity resides.
     """
 
+    pos = Vector(0, 0)
+    game_map = None
+
     @abstractmethod
-    def __init__(self, name, pos, type, char, color, blocks, render_priority, game_map):
+    def __init__(self, id, name, type, char, color, blocks, render_priority):
+        self.id = id
```
Comment (Elkasitu) [inline entities.py]: same

Comment (Elkasitu) [reply entities.py]: you should also add it to the docstring

File: entities.py [hunk 3]
```diff
@@ -101,6 +99,24 @@ def move_towards(self, dest):
         if self.game_map.walkable[next_tile] and not self.game_map.get_blocking_entity_at_location(next_tile):
             self.move(direction)
 
+    def place(self, game_map, position):
```
Comment (Elkasitu) [inline entities.py]: I don't know how I feel about this, IMO this should be either in the registry or in the dungeon code

File: entities.py [hunk 4]
```diff
@@ -114,8 +130,8 @@ class Item(Entity):
         weight (float): Weight of the item, defaults to 1.0
     """
 
-    def __init__(self, name, pos, char, color, game_map, blocks=False, effect=None, weight=1.0):
-        super().__init__(name, pos, 'item', char, color, blocks, RenderPriority.ITEM, game_map)
+    def __init__(self, id, name, char, color, blocks=False, effect=None, weight=1.0):
+        super().__init__(id, name, 'item', char, color, blocks, RenderPriority.ITEM)
```
Comment (Elkasitu) [inline entities.py]: same

File: entities.py [hunk 5]
```diff
@@ -163,7 +181,7 @@ def __init__(self, name, pos, behavior, char, color, game_map):
         self._gold = 100
         # technically all actors will have a backpack, but since it's a singleton
         # and only player has access to it, it doesn't matter.
-        self.backpack = Backpack()
+        self.backpack = Backpack(registry)
```
Comment (Elkasitu) [inline entities.py]: I think this should be the other way around, we access the backpack through the registry

File: entities.py (hunk 3 above)
Comment (monzo94) [reply entities.py]: OK, I'll put it in the dungeon then which also makes sense. Having the registry handle entity placement doesn't make much sense to me.

**Pull request merged**